core/src/main/java/com/linecorp/armeria/client/ClientBuilder.java

```diff
@@ -84,6 +159,38 @@ public ClientBuilder factory(ClientFactory factory) {
      */
     public <T> T build(Class<T> clientType) {
         requireNonNull(clientType, "clientType");
-        return factory.newClient(uri, clientType, buildOptions());
+        checkArguments();
+
+        if (uri != null) {
+            return factory.newClient(uri, clientType, buildOptions());
+        } else if (path != null) {
+            return factory.newClient(scheme(), endpoint, path, clientType, buildOptions());
+        } else {
+            return factory.newClient(scheme(), endpoint, clientType, buildOptions());
+        }
+    }
+
+    private Scheme scheme() {
+        return scheme == null ? Scheme.of(format, requireNonNull(protocol)) : scheme;
+    }
+
+    private void checkArguments() {
+        if (uri != null) {
+            return;
+        }
+
+        if (endpoint == null) {
+            throw new IllegalStateException("both uri and endpoint are not given");
+        }
+
+        if (scheme == null && protocol == null) {
+            throw new IllegalStateException("both scheme and protocol are not given");
+        }
+    }
```

Comment:
I think we do not need this method since the fields are all final and we already did null check in the constructor. And there's no case that the condition violates this states.

Reply:
Let me remove this checking method.
I think constructors with required parameters are enough.